vicodyn/src/vicodyn/proxy.cpp

```diff
@@ -410,24 +480,24 @@ class vicodyn_dispatch_t : public std::enable_shared_from_this<vicodyn_dispatch_
     auto retry_unsafe() -> void {
         COCAINE_LOG_INFO(endpoint.logger, "retrying");
         request_context->register_retry();
-        if(backward_stream.started()) {
+        if (backward_stream.started()) {
             throw error_t("retry is forbidden - response chunk was sent");
         }
-        if(request_context->retry_count() > proxy.balancer->retry_count()) {
-            throw error_t("maximum number of retries reached");
+        if (buffer.failed()) {
+            throw error_t("retry is forbidden - client sent error");
+        }
+        if (buffer.overfull()) {
+            throw error_t("retry is forbidden - buffer overflow");
+        }
+        if (request_context->retry_count() > proxy.balancer->retry_count()) {
+            throw error_t("retry is forbidden - maximum number of retries reached");
         }
         request_context->add_checkpoint("retry");
 
-        choose_endpoint();
+        choose_endpoint(buffer.headers(), buffer.event());
 
-        for(size_t i = 0; i < buffer.chunks.size(); i++) {
-            endpoint.forward_stream.chunk(buffer.chunk_headers[i], buffer.chunks[i]);
-        }
-        if (buffer.choke) {
-            endpoint.forward_stream.close(buffer.choke.get());
-        } else if (buffer.error) {
-            endpoint.forward_stream.error(buffer.error->headers, buffer.error->ec, buffer.error->msg);
```

Comment:
In what direction is `forward_stream`? Is it from client to Vicodyn?
Does it mean that in new variant client don't have additional info about error type?